python/tests/kat/t_active_health_check.py

```diff
@@ -0,0 +1,140 @@
+        # that one of the pods is misbehaving before we start blasting requests out.
+        yield Query(self.url("healthcheck/"), expected=[200, 500], phase=2)
+        yield Query(self.url("healthcheck/"), expected=[200, 500], phase=3)
+
+        # Make 1000 requests split into two groups to reduce any flakes
+        for _ in range(500):
+            yield Query(self.url("healthcheck/"), expected=[200, 500], phase=4)
+            time.sleep(0.06)
+
+        for _ in range(500):
+            yield Query(self.url("healthcheck/"), expected=[200, 500], phase=5)
+            time.sleep(0.06)
+
+    def check(self):
+        print(self.results)
+        # Check the manual requests first
+        assert self.results[0].status == 200
+        assert self.results[1].status == 200
+        assert self.results[2].status == 200
+
+        # Add up the number of 500 and 200 responses that we got.
+        valid = 0
+        errors = 0
+        for i in range(5, 1005):
+            assert (self.results[i].status == 200 or self.results[i].status == 500) == True
+            if self.results[i].status == 200:
+                valid += 1
+            elif self.results[i].status == 500:
+                errors += 1
+
+        # with 1000 requests and 1/5 being an error response, we should have the following distribution +/- 10
+        # assert 190 <= errors <= 210
+        # assert 790 <= valid <= 810
+
+        # But since we configure health chekcing we should actually see 0 errors because the health checks noticed
+        # that one of the pods was unhealthy and didn't route any traffic to it.
+        msg = "Errors: {}, Valid: {}".format(errors, valid)
+        assert errors == 0, msg
+        assert valid == 1000, msg
+
+
+class NoHealthCheckTest(AmbassadorTest):
+    def init(self):
+        self.target = HealthCheckServer()
+
+    def config(self) -> Generator[Union[str, Tuple[Node, str]], None, None]:
+        yield self, self.format(
+            """
+---
+apiVersion: getambassador.io/v3alpha1
+kind: Mapping
+name:  {self.target.path.k8s}-health
+hostname: "*"
+prefix: /healthcheck/
+service: {self.target.path.fqdn}
+resolver: endpoint
+load_balancer:
+  policy: round_robin
+"""
+        )  # The round robin load balancer is not necessary for the test but should help make the request distribution even across the pods
+
+    def queries(self):
+        yield Query(self.url("healthcheck/"))  # Just making sure things are running
+        yield Query(self.url("ambassador/v0/diag/"))
+
+        yield Query(
+            self.url("healthcheck/makeUnhealthy/")
+        )  # the deployment has 5 pods. This will make one of them start returning errors
+
+        # Make 1000 requests over the course of a minute
+        for _ in range(1000):
+            yield Query(self.url("healthcheck/"), expected=[200, 500])
+
+    def check(self):
+        print(self.results)
+        # Check the manual requests first
+        assert self.results[0].status == 200
```

Comment:
Again, I think even when you don't set an expected if it was to return say a 403 randomly then the Query fails expectation and the check function is never run. At least this is what I observed when adding the other tests.

Reply:
Hmm weird, from all of my tests it will fail the test immediately when it receives an unexpected response from the queries or fails the asserts.

Reply:
Misinterpreted what you meant there for a minute but yep I agree.